src/wp-includes/class-wp-speculation-rules.php

```diff
@@ -0,0 +1,218 @@
+<?php
+/**
+ * Class 'WP_Speculation_Rules'.
+ *
+ * @package WordPress
+ * @subpackage Speculative Loading
+ * @since 6.8.0
+ */
+
+/**
+ * Class representing a set of speculation rules.
+ *
+ * @since 6.8.0
+ * @access private
+ */
+class WP_Speculation_Rules implements JsonSerializable {
+
+	/**
+	 * Stored rules, as a map of `$mode => $rules` pairs.
+	 *
+	 * Every `$rules` value is a map of `$id => $rule` pairs.
+	 *
+	 * @since 6.8.0
+	 * @var array<string, array<string, mixed>>
+	 */
+	private $rules_by_mode = array();
+
+	/**
+	 * Adds a speculation rule to the speculation rules to consider.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @param string               $mode Speculative loading mode. Either 'prefetch' or 'prerender'.
+	 * @param string               $id   Unique string identifier for the speculation rule.
+	 * @param array<string, mixed> $rule Associative array of rule arguments.
+	 * @return bool True on success, false if invalid parameters are provided.
+	 */
+	public function add_rule( string $mode, string $id, array $rule ): bool {
+		if ( ! wp_is_valid_speculation_rules_mode( $mode ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				sprintf(
+					/* translators: %s: invalid mode value */
+					__( 'The value "%s" is not a valid speculation rules mode.' ),
+					esc_html( $mode )
+				),
+				'6.8.0'
+			);
+			return false;
+		}
+
+		if ( ! $this->is_valid_id( $id ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				sprintf(
+					/* translators: %s: invalid ID value */
+					__( 'The value "%s" is not a valid ID for a speculation rule.' ),
+					esc_html( $id )
+				),
+				'6.8.0'
+			);
+			return false;
+		}
```

Comment:
Is this needed? The ID is not exposed anywhere, so it can be any string which can be an array key.

Reply:
I think some kind of validation should be present. All kinds of WordPress identifiers or slugs require specific formats, consider e.g. post types being `wp_template_part` instead of `WP Template Part`. The latter would work too for an array identifier, but still is not considered a proper identifier based on what identifiers are supposed to look like across WordPress.